swupd/delta.go

```diff
@@ -0,0 +1,179 @@
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to create deltas list %s", manifest)
+	}
+
+	const maxRoutines = 3
+	var deltaQueue = make(chan *File)
+	var failedChan = make(chan *File)
+	var failedList = make(chan []*File)
+	var wg sync.WaitGroup
+	wg.Add(maxRoutines)
+
+	// Start a collector for the failure cases
+	go func() {
+		var failed []*File
+		// Drains the failedChan appropriately
+		for f := range failedChan {
+			failed = append(failed, f)
+		}
+		failedList <- failed // This list is what we return in the end
+	}()
+
+	// Don't flood the system with goroutines, delta creation takes up an
+	// incredibly amount of memory, so we cannot max out on goroutines
+	for i := 0; i < maxRoutines; i++ {
+		go func(statedir string) {
+			defer wg.Done()
+
+			// Take in jobs from the queue and try to make a delta
+			for f := range deltaQueue {
+				err := createDelta(c, f, statedir, f.DeltaPeer.Version, f.DeltaPeer.Hash)
+				if err != nil {
+					failedChan <- f
+				}
+			}
+		}(statedir)
+	}
+
+	// Send jobs to the queue for delta goroutines to pick up
+	for _, file := range deltas {
+		deltaQueue <- file
+	}
+
+	// Send message that no more jobs are being sent
+	close(deltaQueue)
+	wg.Wait()
+
+	// Once wait finishes we can signal the collector that no more jobs exist
+	close(failedChan)
+
+	return <-failedList, nil
+}
+
+func consolidateDeltaFiles(manifest *Manifest, from uint32, c config) ([]*File, error) {
+	var files []*File
+
+	if manifest == nil {
+		return nil, nil
+	}
+
+	for _, file := range manifest.Files {
+		if file.Version <= from {
+			continue
+		}
+		if file.DeltaPeer == nil {
+			continue
+		}
+		if file.Type != TypeFile ||
+			file.DeltaPeer.Type != TypeFile {
+			continue
+		}
+
+		deltadir := filepath.Join(c.outputDir, fmt.Sprintf("%d", file.Version), "delta")
+		fromToString := fmt.Sprintf("%d-%d-%s-%s", file.DeltaPeer.Version, file.Version, file.DeltaPeer.Hash, file.Hash)
+
+		deltaFile := filepath.Join(deltadir, fromToString)
+
+		// Only add to list if the delta file does not already exist on the system
+		if _, err := os.Stat(deltaFile); os.IsNotExist(err) {
+			// Only create deltas for files bigger than 200 bytes
+			fullname := filepath.Join(c.imageBase, fmt.Sprintf("%d", file.Version), "full"+file.Name)
+			file.Info, _ = os.Stat(fullname)
+			if file.Info.Size() > 200 {
+				files = append(files, file)
+			}
+		}
+
+	}
+
+	return files, nil
+}
+
+func createDelta(c config, file *File, statedir string, fromVersion uint32, fromHash Hashval) error {
+	// File without peers cannot have deltas
+	if file.DeltaPeer == nil {
+		return nil
+	}
+
+	// We only support deltas between two regular files for now
+	// TODO: Support directory -> file in the future, this is a complex case
+	if file.Type != TypeFile || file.DeltaPeer.Type != TypeFile {
+		return nil
+	}
+
+	newfile := filepath.Join(c.imageBase, fmt.Sprintf("%d", file.Version), "full"+file.Name)
```

Comment:
Ohhh, I get it now. I was having a heck of a time figuring out what `"full"+file.Name` was doing because I forgot that `file.Name` would have a leading `/`. This is precisely why `filepath.Join` is useful. Please change that last argument to `(..., "full", file.Name)`